File: sdk-python/agentmonitor/prompt_runtime.py
"""PR-12: Prompt Runtime 客户端。

对应后端 ``GET /api/v2/runtime/prompts/:name`` 接口，提供：

- ETag / 304 协商缓存
- 硬 TTL 缓存（默认 60 秒）
- 线程安全（与 Python SDK 其它模块一致，使用 threading.Lock 保护内存缓存）
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from typing import Any, Dict, Literal, Optional

try:
    import requests
except ImportError:  # pragma: no cover - requests 是必选依赖
    raise ImportError("请安装 requests: pip install requests")


PromptEnvironment = Literal["production", "staging", "development"]


@dataclass
class ResolvedPrompt:
    """Runtime 解析返回的已发布 Prompt。"""

    prompt_id: str
    prompt_name: str
    prompt_version_id: str
    version_number: int
    environment: str
    content: str
    variables_schema: Optional[Dict[str, Any]] = None
    model_defaults: Optional[Dict[str, Any]] = None
    etag: str = ""
    deployed_at: str = ""

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "ResolvedPrompt":
        """从后端 JSON 构造（后端字段为 camelCase）。"""
        return cls(
            prompt_id=data["promptId"],
            prompt_name=data["promptName"],
            prompt_version_id=data["promptVersionId"],
            version_number=data["versionNumber"],
            environment=data["environment"],
            content=data["content"],
            variables_schema=data.get("variablesSchema"),
            model_defaults=data.get("modelDefaults"),
            etag=data.get("etag", ""),
            deployed_at=data.get("deployedAt", ""),
        )


@dataclass
class _CacheEntry:
    prompt: ResolvedPrompt
    cached_at: float


class PromptRuntimeClient:
    """Prompt Runtime 客户端。"""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        project_id: str,
        default_max_age_sec: float = 60.0,
        session: Optional[requests.Session] = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._project_id = project_id
        self._default_max_age = default_max_age_sec
        self._cache: Dict[str, _CacheEntry] = {}
        self._lock = threading.Lock()
        self._session = session or requests.Session()

    def _cache_key(self, prompt_name: str, environment: str) -> str:
        return f"{self._project_id}::{environment}::{prompt_name}"

    def get(
        self,
        prompt_name: str,
        environment: PromptEnvironment = "production",
        max_age_sec: Optional[float] = None,
        force_refresh: bool = False,
        timeout: float = 10.0,
    ) -> ResolvedPrompt:
        """按名称解析一个已发布的 Prompt。

        - 未过期缓存直接命中；
        - 过期后带 If-None-Match 协商；
        - 304 则延长缓存寿命；200 覆盖；非 2xx/304 抛异常。
        """
        ttl = self._default_max_age if max_age_sec is None else max_age_sec
        key = self._cache_key(prompt_name, environment)
        now = time.monotonic()

        with self._lock:
            existing = self._cache.get(key)
        if not force_refresh and existing is not None and now - existing.cached_at < ttl:
            return existing.prompt

        url = (
            f"{self._base_url}/api/v2/runtime/prompts/"
            f"{requests.utils.quote(prompt_name, safe='')}"
        )
        params = {"projectId": self._project_id, "environment": environment}
        headers = {"Authorization": f"Bearer {self._api_key}"}
        if existing is not None:
            headers["If-None-Match"] = existing.prompt.etag

        resp = self._session.get(url, params=params, headers=headers, timeout=timeout)
        if resp.status_code == 304 and existing is not None:
            with self._lock:
                self._cache[key] = _CacheEntry(prompt=existing.prompt, cached_at=time.monotonic())
            return existing.prompt

        if not resp.ok:
            raise RuntimeError(
                f"Prompt runtime {resp.status_code}: {resp.text.strip() or resp.reason}"
            )

        data = resp.json()
        prompt = ResolvedPrompt.from_dict(data)
        with self._lock:
            self._cache[key] = _CacheEntry(prompt=prompt, cached_at=time.monotonic())
        return prompt

    def clear_cache(self) -> None:
        """清空全部缓存。"""
        with self._lock:
            self._cache.clear()

    @property
    def cache_size(self) -> int:
        """当前缓存条目数（主要用于测试 / 可观测性）。"""
        with self._lock:
            return len(self._cache)


__all__ = [
    "PromptRuntimeClient",
    "ResolvedPrompt",
    "PromptEnvironment",
]

File: sdk/src/prompt-runtime.ts
/**
 * PR-12 Prompt Runtime 客户端
 *
 * 对应后端 GET /api/v2/runtime/prompts/:name?environment= 接口。
 * - 内置 ETag/304 缓存，命中 304 时不反序列化 body
 * - 支持 maxAgeMs 硬 TTL，避免高频请求下重复打后端
 * - 缓存键：projectId + promptName + environment
 */

export interface ResolvedPrompt {
  promptId: string;
  promptName: string;
  promptVersionId: string;
  versionNumber: number;
  environment: string;
  content: string;
  variablesSchema: Record<string, unknown> | null;
  modelDefaults: Record<string, unknown> | null;
  etag: string;
  deployedAt: string;
}

export interface GetPromptOptions {
  /** 运行环境，默认 production */
  environment?: 'production' | 'staging' | 'development';
  /** 硬 TTL（毫秒），默认 60_000；在 TTL 内直接复用缓存，不发请求 */
  maxAgeMs?: number;
  /** 强制跳过缓存（仍会带 If-None-Match 走 304 协商） */
  forceRefresh?: boolean;
  /** AbortSignal，用于调用方主动取消 */
  signal?: AbortSignal;
}

interface CacheEntry {
  prompt: ResolvedPrompt;
  cachedAt: number;
}

export interface PromptRuntimeClientOptions {
  baseUrl: string;
  apiKey: string;
  projectId: string;
  /** 默认 TTL，毫秒 */
  defaultMaxAgeMs?: number;
  /** 自定义 fetch（主要用于测试） */
  fetchImpl?: typeof fetch;
}

/**
 * Prompt Runtime 客户端。线程安全（JS 单线程模型下 Map 操作原子）。
 */
export class PromptRuntimeClient {
  private readonly cache = new Map<string, CacheEntry>();
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly projectId: string;
  private readonly defaultMaxAgeMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: PromptRuntimeClientOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/$/, '');
    this.apiKey = opts.apiKey;
    this.projectId = opts.projectId;
    this.defaultMaxAgeMs = opts.defaultMaxAgeMs ?? 60_000;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  /**
   * 构造缓存键。
   */
  private cacheKey(promptName: string, environment: string): string {
    return `${this.projectId}::${environment}::${promptName}`;
  }

  /**
   * 按名称解析一个已发布的 Prompt。
   *
   * 行为：
   * 1. 若有未过期缓存且未 forceRefresh，直接返回缓存
   * 2. 否则发请求，带 If-None-Match；命中 304 时延长缓存寿命
   * 3. 200 时写入新缓存
   *
   * @param promptName - Prompt 名称
   * @param options - 可选环境/TTL/强制刷新/AbortSignal
   * @returns ResolvedPrompt，未发布时抛 404 错误
   */
  async get(promptName: string, options: GetPromptOptions = {}): Promise<ResolvedPrompt> {
    const environment = options.environment ?? 'production';
    const maxAgeMs = options.maxAgeMs ?? this.defaultMaxAgeMs;
    const key = this.cacheKey(promptName, environment);
    const existing = this.cache.get(key);
    const now = Date.now();

    if (!options.forceRefresh && existing && now - existing.cachedAt < maxAgeMs) {
      return existing.prompt;
    }

    const query = new URLSearchParams({ projectId: this.projectId, environment });
    const url = `${this.baseUrl}/api/v2/runtime/prompts/${encodeURIComponent(promptName)}?${query.toString()}`;
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
    };
    if (existing) {
      headers['If-None-Match'] = existing.prompt.etag;
    }

    const res = await this.fetchImpl(url, { method: 'GET', headers, signal: options.signal });

    if (res.status === 304 && existing) {
      // 延长 TTL，复用旧内容
      this.cache.set(key, { prompt: existing.prompt, cachedAt: now });
      return existing.prompt;
    }

    if (!res.ok) {
      const body = await res.text().catch(() => '');
      throw new Error(`Prompt runtime ${res.status}: ${body || res.statusText}`);
    }

    const data = (await res.json()) as ResolvedPrompt;
    this.cache.set(key, { prompt: data, cachedAt: now });
    return data;
  }

  /**
   * 清空全部缓存（测试 / 紧急切换环境时使用）。
   */
  clearCache(): void {
    this.cache.clear();
  }

  /**
   * 当前缓存条目数（可观测性）。
   */
  get cacheSize(): number {
    return this.cache.size;
  }
}
